PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
+        """
+        Remove the passed modules and hence all hook scripts cached for these
+        modules from this cache.
+
+        Parameters
+        ----------
+        module_names : list
+            List of all fully-qualified module names to be removed.
+        """
+
+        for module_name in set(module_names):  # Eliminate duplicate entries.
+            if module_name in self:
+                del self[module_name]
+
+
+class ModuleHook(object):
+    """
+    Cached object encapsulating a lazy loadable hook script.
+
+    This object's `code` property provides the currently loaded contents (e.g.,
+    attributes, functions) of this hook script. The first access of this
+    property silently lazily loads this script into an in-memory code object.
+
+    Parameters
+    ----------
+    module_graph : ModuleGraph
+        Current module graph.
+    module_name : str
+        Name of the module hooked by this hook script.
+    hook_filename : str
+        Absolute or relative path of this hook script.
+    hook_module_name : str
+        Name of the in-memory module of this hook script's interpreted contents.
+    _scope : module
+        In-memory module of this hook script's interpreted contents, lazily
+        loaded on the first access of the `scope` property _or_ `None` if this
+        property has yet to be accessed.
+    """
+
+    ## Private methods
+
+    def __init__(
+        self,
+        module_graph,
+        module_name,
+        hook_filename,
+        hook_module_name_prefix,
+    ):
+        """
+        Initialize this metadata.
+
+        Parameters
+        ----------
+        module_graph : ModuleGraph
+            Current module graph.
+        module_name : str
+            Name of the module hooked by this hook script.
+        hook_filename : str
+            Absolute or relative path of this hook script.
+        hook_module_name_prefix : str
+            String prefixing the name of the in-memory module for this hook
+            script's code object. To avoid namespace clashes with similar
+            modules created by other `ModuleHook` objects in other
+            `ModuleHookCache` containers, this string _must_ be unique to the
+            `ModuleHookCache` container containing this `ModuleHook` object. If
+            this string is non-unique, an existing in-memory module will be
+            erroneously reused when lazily loading this hook script, thus
+            erroneously resanitizing previously sanitized hook script attributes
+            (e.g., `datas`) with the `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
+        # avoiding circular reference issues. See ModuleHookCache.__init__().
+        self.module_graph = module_graph
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
```

Comment:
Please be more verbose on why we need this. Eg.
> Each import of a hook must get it's own fresh scope, otherwise artefacts from the last run may change the results. So the hooks gets a new name on each import, so the import machinery will not reuse the existing scope.

Assuming, this is the reason, of course.

OTOH 1: Is "otherwise artefacts from the last run may change the results" correct at all? It may only be relevant when having several `Analysis()` in one .spec-file (e.g. for MERGE) and in the test-suite. But what could change between two hook runs? Of course the graph may change due to e.g. excluded imports or different pathex.

OTOH 2: Couldn't "artefacts from the last run" be even beneficial, since we could e.g. cache results from `collect_submodules` or from the even more expensive evaluations?

My sense says we'd better assume "otherwise artefacts from the last run may change the results" is correct.